timely_beliefs/beliefs/utils.py

```diff
@@ -1030,3 +1036,66 @@ def meta_repr(
             for attr in tb_structure._metadata
         ]
     )
+
+
+def upsample_beliefs_data_frame(
+    df: "classes.BeliefsDataFrame" | pd.DataFrame, event_resolution: timedelta, keep_nan_values: bool = False
+) -> "classes.BeliefsDataFrame":
+    """Because simply doing df.resample().ffill() does not correctly resample the last event in the data frame.
+
+    :param df:                  In case of a regular pd.DataFrame, make sure to set df.event_resolution before passing it to this function.
+    :param event_resolution:    Resolution to upsample to.
+    :param keep_nan_values:     If True, place back resampled NaN values. Drops NaN values by default.
+    """
+    from_event_resolution = df.event_resolution
+    resample_ratio = pd.to_timedelta(
+        to_offset(from_event_resolution)
+    ) / pd.Timedelta(event_resolution)
+    if keep_nan_values:
+        # back up NaN values
+        unique_event_value_not_in_df = df["event_value"].abs().sum() + 1
+        df = df.fillna(unique_event_value_not_in_df)
+    new_index = initialize_index(
+        start=df.index[0],
+        end=df.index[-1] + from_event_resolution,
+        resolution=event_resolution,
+    )
+    # Reindex to introduce NaN values, then forward fill by the number of steps
+    # needed to have the new resolution cover the old resolution.
+    # For example, when resampling from a resolution of 30 to 20 minutes (NB frequency is 1 hour):
+    # event_start               event_value
+    # 2020-03-29 10:00:00+02:00 1000.0
+    # 2020-03-29 11:00:00+02:00 NaN
+    # 2020-03-29 12:00:00+02:00 2000.0
+    # After reindexing
+    # event_start               event_value
+    # 2020-03-29 10:00:00+02:00 1000.0
+    # 2020-03-29 10:20:00+02:00 NaN
+    # 2020-03-29 10:40:00+02:00 NaN
+    # 2020-03-29 11:00:00+02:00 NaN
+    # 2020-03-29 11:20:00+02:00 NaN
+    # 2020-03-29 11:40:00+02:00 NaN
+    # 2020-03-29 12:00:00+02:00 2000.0
+    # 2020-03-29 12:20:00+02:00 NaN
+    # 2020-03-29 12:40:00+02:00 NaN
+    # After filling a limited number of NaN values (ceil(30/20)-1 == 1)
+    # event_start               event_value
+    # 2020-03-29 10:00:00+02:00 1000.0
+    # 2020-03-29 10:20:00+02:00 1000.0
+    # 2020-03-29 10:40:00+02:00 NaN
+    # 2020-03-29 11:00:00+02:00 NaN
+    # 2020-03-29 11:20:00+02:00 NaN
+    # 2020-03-29 11:40:00+02:00 NaN
+    # 2020-03-29 12:00:00+02:00 2000.0
+    # 2020-03-29 12:20:00+02:00 2000.0
+    # 2020-03-29 12:40:00+02:00 NaN
+    df = df.reindex(new_index).fillna(
+        method="pad",
```

Comment:
The method `pad` is being replaced by `ffill` in an upcoming Panda version:

https://stackoverflow.com/a/48586122

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.fillna.html